polkadot/runtime/parachains/src/paras/mod.rs

```diff
@@ -1160,59 +1188,149 @@ pub mod pallet {
 			MostRecentContext::<T>::insert(&para, context);
 			Ok(())
 		}
+
+		/// Sets the storage for the authorized current code hash of the parachain.
+		/// If not applied, it will be removed at the `System::block_number() + valid_period` block.
+		///
+		/// This can be useful, when triggering `Paras::force_set_current_code(para, code)`
+		/// from a different chain than the one where the `Paras` pallet is deployed.
+		///
+		/// The main purpose is to avoid transferring the entire `code` Wasm blob between chains.
+		/// Instead, we authorize `code_hash` with `root`, which can later be applied by
+		/// `Paras::apply_authorized_force_set_current_code(para, code)` by anyone.
+		///
+		/// Authorizations are stored in an **overwriting manner**.
+		#[pallet::call_index(9)]
+		#[pallet::weight(<T as Config>::WeightInfo::authorize_force_set_current_code_hash())]
+		pub fn authorize_force_set_current_code_hash(
+			origin: OriginFor<T>,
+			para: ParaId,
+			new_code_hash: ValidationCodeHash,
+			valid_period: BlockNumberFor<T>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			let now = frame_system::Pallet::<T>::block_number();
+			let expire_at = now.saturating_add(valid_period);
+
+			// insert authorized code hash and make sure to overwrite existing one for a para.
+			AuthorizedCodeHash::<T>::insert(&para, (new_code_hash, expire_at));
+			Self::deposit_event(Event::CodeAuthorized {
+				para_id: para,
+				code_hash: new_code_hash,
+				expire_at,
+			});
+
+			Ok(())
+		}
+
+		/// Applies the already authorized current code for the parachain,
+		/// triggering the same functionality as `force_set_current_code`.
+		#[pallet::call_index(10)]
+		#[pallet::weight(<T as Config>::WeightInfo::apply_authorized_force_set_current_code(new_code.0.len() as u32))]
+		pub fn apply_authorized_force_set_current_code(
+			_origin: OriginFor<T>,
+			para: ParaId,
+			new_code: ValidationCode,
+		) -> DispatchResultWithPostInfo {
+			// no need to ensure, anybody can do this
+
+			// Ensure `new_code` is authorized
+			let _ = Self::validate_code_is_authorized(&new_code, &para)?;
+			// Remove authorization
+			AuthorizedCodeHash::<T>::remove(para);
+
+			// apply/dispatch
+			Self::do_force_set_current_code_update(para, new_code);
+
+			// if ok, then allows "for free"
+			let post = PostDispatchInfo {
+				// consume the rest of the block to prevent further transactions
+				actual_weight: Some(T::BlockWeights::get().max_block),
```

Comment:
Why do we need to do this? I suppose the code blob can be large, but ref-time wise this doesn't seem like a large transaction at all.

Reply:
Good question! Honestly, I'm not entirely sure. I just followed this pattern: `authorize + apply_authorized` from the `frame_system` pallet, which is how we upgrade runtimes: https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/system/src/lib.rs#L887-L891.